include/boost/url/impl/url_base.ipp

```diff
@@ -1744,6 +1751,287 @@ operator[](string_view key) const
+    pct.encode(v.s_ + pos.off_ + 1, sk.data());
+
+    // Value
+    if( nv )
+    {
+        v.s_[pos.off_ + 1 + nk] = '=';
+        pct.encode(v.s_ + pos.off_ + 1 + nk + 1, sv.data());
+    }
+    BOOST_ASSERT(pos.off_ + n <= v.pt_.offset[detail::id_frag]);
+
+    // Deal with delimiters
+    if( pos.off_ != v.pt_.offset[detail::id_query] )
+    {
+        v.s_[pos.off_] = '&';
+    }
+    else
+    {
+        v.s_[pos.off_] = '?';
+        if( pos.off_ + n != v.pt_.offset[detail::id_frag] )
+            v.s_[pos.off_ + n] = '&';
+    }
+
+    // Return a valid iterator.
+    pos.parse();
+    return pos;
+}
+
+auto
+url_base::
+params_type::
+insert_encoded( iterator pos, value_type p ) ->
+    iterator
+{
+    auto const pct = detail::pchar_pct_set();
+    pct.validate(p.k_);
+    pct.validate(p.v_);
+    return insert_encoded_impl(pos, p.k_, p.v_);
+}
+
+auto
+url_base::
+params_type::
+insert( iterator pos, value_type p ) ->
+    iterator
+{
+    auto const pct = detail::pchar_pct_set();
+    auto const nk = pct.encoded_size(p.k_);
+    auto const nv = pct.encoded_size(p.v_);
+    return insert_impl(pos, p.k_, nk, p.v_, nv);
+}
+
+auto
+url_base::
+params_type::
+replace_encoded( iterator pos, value_type p ) ->
+    iterator
+{
+    (void) detail::pchar_pct_set().validate(p.k_);
+    (void) detail::pchar_pct_set().validate(p.v_);
+    BOOST_ASSERT(v_ != nullptr);
+    url_base & v = *v_;
+    BOOST_ASSERT(v.size() == v.a_.size());
+
+    auto const nk = p.k_.size();
+    auto const nv = p.v_.size();
+    auto const n0 = pos.nk_ + pos.nv_;
+    auto const n = 1 + nk + (nv!=0) + nv; // Delimiter, key, =, value
+
+    // Reserve just enough memory for the replacement
+    if( n0 < n )
+        v.s_ = v.a_.reserve(v.a_.size() + n - n0);
+
+    // Erase, then insert
+    auto const cap = v.a_.capacity();
+    auto r = insert_encoded_impl(
+        erase(pos),
+        p.k_,
+        p.v_);
+
+    // If the reserve was exceeded, we have a strong guarantee violation
+    BOOST_ASSERT(v.a_.capacity() == cap);
+    (void) cap;
+
+    return r;
+}
+
+auto
+url_base::
+params_type::
+replace( iterator pos, value_type p ) ->
+    iterator
+{
+    BOOST_ASSERT(v_ != nullptr);
+    url_base & v = *v_;
+    BOOST_ASSERT(v.size() == v.a_.size());
+
+    auto const pct = detail::pchar_pct_set();
+    auto const nk = pct.encoded_size(p.k_);
+    auto const nv = pct.encoded_size(p.v_);
+    auto const n0 = pos.nk_ + pos.nv_;
+    auto const n = 1 + nk + (nv!=0) + nv; // Delimiter, key, =, value
+
+    // Reserve just enough memory for the replacement
+    if( n0 < n )
+        v.s_ = v.a_.reserve(v.a_.size() + n - n0);
+
+    // Erase, then insert
+    auto const cap = v.a_.capacity();
+    auto r = insert_impl(
+        erase(pos),
+        p.k_,
+        nk,
+        p.v_,
+        nv);
+
+    // If the reserve was exceeded, we have a strong guarantee violation
+    BOOST_ASSERT(v.a_.capacity() == cap);
+    (void) cap;
```

Comment:
I think the verbosity and awkward code flow stems from the fact that we are calling a throwing function `insert_impl` from other functions which have effects and postconditions (e.g. strong guarantee).

We could simply make it a requirement that the caller reserves the necessary space, and then `insert_impl` would merely assert that the capacity is sufficient. Then we won't need to have these local variables which are not used in release builds and we won't need these unnecessary asserts.